theano/contrib/mkl/mkl_type.py

```diff
@@ -0,0 +1,350 @@
+            Py_INCREF(%(name)s);
+        }
+        """ % locals(), file=sio)
+
+        return sio.getvalue()
+
+    def c_cleanup(self, name, sub):
+        """
+        Add cleanup code here.
+        """
+        return """
+        if (%(name)s) {
+            Py_XDECREF(%(name)s);
+            %(name)s = NULL;
+        }
+        """ % locals()
+
+    def c_sync(self, name, sub):
+        """
+        When the computations are done, transfer the variables from the C
+        structure we put them in to the destination Python object. This will
+        only be called for the outputs.
+        """
+        return """
+        if (NULL == %(name)s) {
+            Py_XDECREF(py_%(name)s);
+            py_%(name)s = Py_None;
+            Py_INCREF(py_%(name)s);
+        } else {
+            if (py_%(name)s != (PyObject*)%(name)s) {
+                Py_XDECREF(py_%(name)s);
+                py_%(name)s = (PyObject*)%(name)s;
+                Py_INCREF(py_%(name)s);
+            }
+            assert(py_%(name)s->ob_refcnt);
+        }
+        """ % locals()
+
+    def c_headers(self):
+        return ['mkl_ndarray.h']
+
+    def c_header_dirs(self):
+        ret = [os.path.dirname(mkl_ndarray.__file__), os.path.dirname(__file__)]
+        return ret
+
+    def c_lib_dirs(self):
+        ret = [os.path.dirname(mkl_ndarray.__file__)]
+        return ret
+
+    def c_libraries(self):
+        return ['mkl_ndarray']
+
+    def c_code_cache_version(self):
+        return (1, 0, 0)
+
+    def c_compile_args(self):
+        return ['-Wl,-R' + os.path.dirname(mkl_ndarray.__file__)]
+
+    def get_shape_info(self, obj):
+        return obj.shape
+
+    def get_size(self, shape_info):
+        """
+        If shape_info is given, return memory size of all data,
+        else size of single element is returned.
+        """
+        if shape_info:
+            return numpy.prod(shape_info) * numpy.dtype(self.dtype).itemsize
+        else:
+            return numpy.dtype(self.dtype).itemsize
+
+
+# expandable_types: can add an extra dimension and for which Scan can deal with.
+# TODO: further check with MKLNdarrayType to support that.
+# theano.compile.ops.expandable_types += (MKLNdarrayType,)
+
+# Register C code for ViewOp on CudaNdarrayType
+theano.compile.register_view_op_c_code(
+    MKLNdarrayType,
+    """
+    Py_XDECREF(%(oname)s);
+    %(oname)s = %(iname)s;
+    Py_XINCREF(%(oname)s);
+    """,
+    version=(1,))
+
+theano.compile.register_shape_c_code(
+    MKLNdarrayType,
+    """
+    npy_intp shape[] = {MKLNdarray_NDIM(%(iname)s)};
+    if (%(oname)s == NULL || (PyArray_DIMS(%(oname)s)[0] != shape[0])) {
+        Py_XDECREF(%(oname)s);
+        %(oname)s = (PyArrayObject*)PyArray_SimpleNew(1, shape, NPY_INT64);
+    }
+
+    for (int i=0; i<shape[0]; i++) {
+        ((npy_int64*)PyArray_GETPTR1(%(oname)s, i))[0] = MKLNdarray_DIMS(%(iname)s)[i];
+    }
+    """,
+    version=(1,))
+
+theano.compile.register_shape_i_c_code(
```

Comment:
ideally, theano.compile.register_deep_copy_op_c_code() and 
theano.compile.register_specify_shape_c_code() should have c code as in tensor/type.py, but this isn't mandatory.